### ISSUE TITLE
warnings in the Arr_2 testsuite about `throw(..)`

### ISSUE THREAD

**Issue by lrineau** [CGAL/cgal#1886]
There are new warnings with a very recent g++ compiler in the Arr_2 testsuite:

```
In file included from /mnt/testsuite/include/CGAL/Algebraic_kernel_d/Curve_analysis_2.h:51:0,
                 from /mnt/testsuite/include/CGAL/Algebraic_kernel_d/Algebraic_curve_kernel_2.h:55,
                 from /mnt/testsuite/include/CGAL/Arr_algebraic_segment_traits_2.h:33,
                 from /home/cgal_tester/build/src/cmake/platforms/Fedora-rawhide-Release/test/Arrangement_on_surface_2/test_geom_traits.h:63,
                 from /home/cgal_tester/build/src/cmake/platforms/Fedora-rawhide-Release/test/Arrangement_on_surface_2/test_traits.cpp:64:
.../include/CGAL/Algebraic_kernel_d/Event_line_builder.h:120:9: warning: dynamic exception specifications are deprecated in C++11; use 'noexcept' instead [-Wdeprecated]
         throw(internal::Zero_resultant_exception<Polynomial_2>)
         ^~~~~
.../include/CGAL/Algebraic_kernel_d/Event_line_builder.h:154:9: warning: dynamic exception specifications are deprecated in C++11; use 'noexcept' instead [-Wdeprecated]
         throw(CGAL::internal::Non_generic_position_exception) {
         ^~~~~
.../include/CGAL/Algebraic_kernel_d/Event_line_builder.h:534:9: warning: dynamic exception specifications are deprecated in C++11; use 'noexcept' instead [-Wdeprecated]
         throw(CGAL::internal::Non_generic_position_exception) {
         ^~~~~
.../include/CGAL/Algebraic_kernel_d/Event_line_builder.h:616:9: warning: dynamic exception specifications are deprecated in C++11; use 'noexcept' instead [-Wdeprecated]
         throw(CGAL::internal::Non_generic_position_exception) {
         ^~~~~
```
https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-4.10-Ic-119/Arrangement_on_surface_2/TestReport_lrineau_Fedora-rawhide-Release.gz

My suggestion is to simply remove those dynamic exception specifications
(the expressions `throw(/*something*/)`): it is well known that it is
inefficient, and that is why it has eventually been deprecated from the C++
norm. That will even be removed from C++17: see http://en.cppreference.com/w/cpp/language/except_spec.

Cc: @CGAL/geometryfactory

**Comment by lrineau** [comment]
The culprit code is actually from AK_2.

**Comment by lrineau** [comment]
Should I do it myself?

Eric, Michael, you are listed as the maintainers of the Ak_d package.

**Comment by lrineau** [comment]
Michael Kerber is the initial author of that file, and Eric is listed as the responsible CGAL editor for him.

Cc also @akobel, who said he is still in the subject of the Ak_d package and things about `<CGAL/Algebraic_kernel_d/Curve_analysis_2.h>`.

**Comment by lrineau** [comment]
I reported that issue three weeks ago! @efifogel, @mhsaar, @akobel, @ericberberich, could you both comment? Who is responsible for that code in CGAL?

**Comment by ericberberich** [comment]
remove or replace with something? isn't there a catch that prevents a full crash if such an error is thrown?

**Comment by lrineau** [comment]
@ericberberich 
Short answer: no, those `throw(something)` specifications are completely useless, for C++11 or later and even for C++98/03. The best solution is just to remove them. That would be different is that was `throw()`.


The longer answer now...

[Dynamic exception specifications][1] (`throw(`*something*`)`) have been replaced in C++11 by [the `noexcept` specifier][2]. The `throw(...)` specification was intended to allow the compiler for better optimizations, but it turned out that it is completely useless:
  - either a function can throw something (and in that case no particular optimization can be done, because the exception throwing is an extra exit point of the function, that cannot be avoided,
  - or the function is specified with `throw()` (deprecated ), `noexcept`, or `noexcept(true)`, and in that case the compiler can remove all exception throwing support from the function, and optimize it further.

What is more there was nothing allowing the compiler to verify the dynamic exception specifications, and the code had to call `std::set_unexcepted` (equivalent to `std::terminate`) if an disallowed exception was thrown. That even adds work to do for compiler.

`throw()` on the other hand is deprecated since C++11, but is considered equivalent to `noexcept`. That one is useful, part of the type system, and verified by the compiler.

[1]: en.cppreference.com/w/cpp/language/except_spec
[2]: http://en.cppreference.com/w/cpp/language/noexcept_spec

**Comment by lrineau** [comment]
Fixed by @afabri in #2016.
